lib/puppet/provider/service/smf.rb

```diff
@@ -135,8 +143,8 @@ def status
       states = service_states
       state = states[1] == "-" ? states[0] : states[1]
     rescue Puppet::ExecutionFailure
-      info "Could not get status on service #{self.name}"
-      return :stopped
+      debug "Could not get status on service #{self.name} #{$!}"
+      return :absent
```

Comment:
@shawnferry - my apologies. I led you astray with asking you to define the `exists?` method. I was mistaken in that it turns out the `service` type isn't actually `ensurable` - which means the `exists?` method (as far as I can tell) won't ever be called/used (unless we use it ourselves as a helper for some other task, and we effectively already have that via a failure of the `service_states` method). The only other example of a service provider defining `exists?` is `launchd` and a quick check seems to indicate that isn't used either. I think by simply changing line 147 to `return :absent` we have what we need to solve the problem of puppet printing `ensure => stopped` for non-existent services, and the `exists?` method can be removed.

This means we actually do still reach this rescue block whenever a service doesn't exist. However, I don't think that's a problem. Puppet won't issue a change in that case, but rather a failure to change. Which is what we want.. I.e., with non-existent service `foo`, with your change to `status` method:

```
-bash-4.1# puppet resource service foo
service { 'foo':
  ensure => 'absent',
}
-bash-4.1# puppet resource service foo ensure=stopped
Error: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Error: /Service[foo]/ensure: change from absent to stopped failed: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
service { 'foo':
  ensure => 'absent',
}
-bash-4.1# puppet apply -e " service { 'foo' : ensure => stopped } " --detailed-exitcodes
Notice: Compiled catalog for solaris11.localdomain in environment production in 0.33 seconds
Error: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Error: /Stage[main]/Main/Service[foo]/ensure: change from absent to stopped failed: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Notice: Applied catalog in 0.08 seconds
-bash-4.1# echo $?
4
```